### ISSUE TITLE
Async / asynchronous execution

### ISSUE THREAD

**Issue by maartenbreddels** [jupyter/nbconvert#1092]
In voila, we run into a performance problem because nbconvert blocks during the execution of cells https://github.com/QuantStack/voila/issues/363. This execution is handled in the tornado handler, causing all other trafic to stop/block. 

Are there plans to have nbconvert use the tornado/zmq/ioloop and have a async way of generating cells results?

**Comment by MSeal** [comment]
Do you mean for the ExecutePreprocessor or any template conversion in general?

From https://github.com/jupyter/nbconvert/issues/1045 we were planning on pulling the execute preprocessor code out of nbconvert, which would be a good opportunity to make it async aware. For other conversions async would probably also be possible in nbconvert 6.0 as we'll be jumping to python 3 only and able to make more dramatic changes in that process.

**Comment by bollwyvl** [comment]
Seems like using tornado's run_on_executor or a pool, as on nbviewer, might
be appropriate for the voila case.

Having said that, async nbconvert is a whacking good idea, and just expect
everything in the pipeline to potentially be async... while still providing
the sync API for a while.

**Comment by maartenbreddels** [comment]
Yes, I plan to open a WIP PR for that on Friday, I have to clean up the code a bit, it's a lot of copy/paste + adding async and await at many places. But I'd rather have a bad PR open to start experimenting/discussing.

**Comment by maartenbreddels** [comment]
> > Seems like using tornado's run_on_executor or a pool, as on nbviewer, might be appropriate for the voila case.
> 
> I think that might be a decent stop-gap measure, @bollwyvl could you link to the code on nbviewer that could give folks a starting place for how to implement this for voila?

@jtpio had a nice PR on that: https://github.com/QuantStack/voila/pull/364/files

But this solution will not work for us since we want to execute cell by cell (async) to allow progressive rendering. What is done in that PR will return the execution after it's done, not after each cell execution.

**Comment by maartenbreddels** [comment]
A POC for an async jupyter client is https://github.com/jupyter/jupyter_client/pull/471
And voila is using it here: https://github.com/QuantStack/voila/pull/374
